src/components/input/input.ts

```diff
@@ -262,8 +272,9 @@ export default class SlInput extends ShoelaceElement implements ShoelaceFormCont
     this.emit('sl-input');
   }
 
-  private handleInvalid() {
+  private handleInvalid(event: Event) {
     this.formControlController.setValidity(false);
+    this.formControlController.emitSlInvalidEvent(event);
```

Comment:
Is it possible to move the emitter inside `setValidity()`? We know it's invalid when we call `setValidity(false)`, but I might be missing a use case for `emitSlInvalidEvent()`. The event could be passed as an optional second param.

Reply:
Actually, that would cause some trouble. See for example the following case, when `this.disabled` switches to `false`. You do not want `sl-invalid` events to be emitted in this case (`sl-invalid` must only be emitted if `(report/check)Validity()` is called):  

```ts
  @watch('disabled', { waitUntilFirstUpdate: true })
  handleDisabledChange() {
    // Disabled form controls are always valid
    this.formControlController.setValidity(this.disabled);
  }
```

By the way, isn't that a bug? I see it in several SL components ... setting `disable` to `true` should result in `setValidity(true)`, yes, but for the other case things are more complex (okay, maybe I just overlook something, maybe there's some implicit `updateValidity` call somehere, but at least it looks suspicious).

If we combine the two functions than in the following overloaded way (method overloading, because `setValidity(true, invalidEvent)` does not make any sense): 

```ts
setValidity(isValid: boolean): void // no `sl-invalid` event will be emitted
setValidity(isValid: false, originalInvalidEvent: Event): void // an `sl-invalid` event will be emitted
```

And this is a bit difficult to understand. By the way it violates the single-responsibilty principle, I guess.

Reply:
Fair enough. Carry on!